api_handler.go

```diff
@@ -12,63 +12,78 @@ import (
 
 	"github.com/jackc/pgx/v5"
 	"github.com/jackc/pgx/v5/pgtype"
-	"github.com/jackc/pgx/v5/pgxpool"
 
 	"github.com/riverqueue/river"
 	"github.com/riverqueue/river/rivertype"
+	"github.com/riverqueue/riverui/internal/apiendpoint"
+	"github.com/riverqueue/riverui/internal/apierror"
 	"github.com/riverqueue/riverui/internal/db"
+	"github.com/riverqueue/riverui/internal/util/dbutil"
 )
 
-type jobCancelRequest struct {
-	JobIDStrings []string `json:"ids"`
-}
-
-type apiHandler struct {
+// A bundle of common utilities needed for many API endpoints.
+type apiBundle struct {
 	client  *river.Client[pgx.Tx]
-	dbPool  *pgxpool.Pool
+	dbPool  DBTXWithBegin
 	logger  *slog.Logger
 	queries *db.Queries
 }
 
-func (a *apiHandler) JobCancel(rw http.ResponseWriter, req *http.Request) {
-	ctx, cancel := context.WithTimeout(req.Context(), 5*time.Second)
-	defer cancel()
+// SetBundle sets all values to the same as the given bundle.
+func (a *apiBundle) SetBundle(bundle *apiBundle) {
+	*a = *bundle
+}
 
-	var payload jobCancelRequest
-	if err := json.NewDecoder(req.Body).Decode(&payload); err != nil {
-		a.logger.ErrorContext(ctx, "error decoding request", slog.String("error", err.Error()))
-		http.Error(rw, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
-		return
-	}
-	jobIDs, err := stringIDsToInt64s(payload.JobIDStrings)
-	if err != nil {
-		a.logger.ErrorContext(ctx, "error decoding job IDs", slog.String("error", err.Error()))
-		http.Error(rw, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
-		return
+// withSetBundle is an interface that's automatically implemented by an types
+// that embed apiBundle. It lets places like tests generically set bundle values
+// on any general endpoint type.
+type withSetBundle interface {
+	// SetBundle sets all values to the same as the given bundle.
+	SetBundle(bundle *apiBundle)
+}
+
+type jobCancelEndpoint struct {
+	apiendpoint.Endpoint[jobCancelRequest, jobCancelResponse]
+	apiBundle
+}
+
+func (*jobCancelEndpoint) Meta() *apiendpoint.EndpointMeta {
+	return &apiendpoint.EndpointMeta{
+		Path:       "POST /api/jobs/cancel",
+		StatusCode: http.StatusOK,
 	}
+}
+
+type jobCancelRequest struct {
+	JobIDs []int64String `json:"ids"`
+}
 
-	updatedJobs := make(map[int64]*rivertype.JobRow)
+type jobCancelResponse struct {
+	Status string `json:"status"`
+}
 
-	if err := pgx.BeginFunc(ctx, a.dbPool, func(tx pgx.Tx) error {
-		for _, jobID := range jobIDs {
+func (a *jobCancelEndpoint) Execute(ctx context.Context, req *jobCancelRequest) (*jobCancelResponse, error) {
```

Comment:
So each endpoint gets an `Execute` function like this one, with the main benefits compared to the alternative being:

* Error handling gets much easier; the handler can just return an error or API error as the second return value and not having to worry about writing the error back or calling `return`.
* Makes general testing much easier because most tests can just send a strongly typed request struct and receive a typed response.
* Moves marshaling/unmarshaling and other common API endpoint features down a layer so the handler code can be made much more succinct.